docs/proposals/20210624-cloneset-scaledown-topology-spread.md

```diff
@@ -0,0 +1,145 @@
+
+1. Node unassigned < assigned
+2. PodPending < PodUnknown < PodRunning
+3. Not ready < ready
+4. Lower pod-deletion cost < higher pod-deletion-cost
+5. Been ready for empty time < less time < more time
+6. Pods with containers with higher restart counts < lower restart counts
+7. Empty creation time pods < newer pods < older pods
+
+What's more, CloneSet should also choose pods to delete according to topology constraints.
+
+## User Stories
+
+1. Pods that on the same node should be deleted preferentially.
+2. Should delete Pods according to topology spread constraints.
+
+## Implementation Details/Notes/Constraints
+
+Add a rank step between (4) and (5). We can now support two Ranker types:
+
+- Ranker for same node spread.
+- Ranker for [Pod Topology Spread Constraints](https://kubernetes.io/docs/concepts/workloads/pods/pod-topology-spread-constraints/).
+
+### SameNodeRanker
+
+SameNodeRanker helps calculate pod ranks according to same node stacking.
+
+- Rank of those pods that are alone on their nodes is `0` by default.
+- For a node that has multiple pods on it, SameNodeRanker will sort these pods by active and give them the index as its rank.
+
+For example:
+
+```
+node-a: pod-0, pod-1
+node-b: pod-2
+node-c: pod-3, pod-4, pod-5
+```
+
+The result will be:
+- Rank of `pod-2` is definitely `0`.
+- Sort pods for `node-a`, if they are `[pod-0, pod-1]` after sorting, then ranks will be `{pod-0: 1, pod-1: 0}`.
+- Sort pods for `node-c`, if they are `[pod-5, pod-3, pod-4]` after sorting, then ranks will be `{pod-5: 2, pod-3: 1, pod-4: 0}`.
+
+### SpreadConstraintsRanker
+
+SpreadConstraintsRanker is a little more complicated than SameNodeRanker.
+
+First, we should define the constraint rule that the ranker should care about:
+
+```golang
+type PodSpreadConstraint struct {
+	TopologyKey     string                   `json:"topologyKey"`
+	SpreadWithRatio bool                     `json:"spreadWithRatio,omitempty"`
+	TopologyRatios  []PodSpreadTopologyRatio `json:"topologyRatios,omitempty"`
+}
+
+type PodSpreadTopologyRatio struct {
+	TopologyValue string `json:"topologyValue,omitempty"`
+	Ratio         int32  `json:"ratio,omitempty"`
+}
+```
+
+- `TopologyKey` is the key of node labels. Nodes that have a label with this key and identical values are considered to be in the same topology.
+- `SpreadWithRatio` indicates whether the different topology values have different spread ratio.
+  By default, all values of this topology key should have the same ratio `1`.
+- `TopologyRatios` defines the ratio of specific values if `SpreadWithRatio` is `true`.
+
+*Note that the constraint rules will be got from CloneSet template (such as `cloneset.spec.template.spec.topologySpreadConstraints`),
+users have no need to configure them explicitly.*
+
+SpreadConstraintsRanker sorting logic:
+1. Split nodes and pods into different values of such topology key.
+2. Calculate the ratio percent for each topology value, it is currentRatio/totalRatio.
```

Comment:
we can simplify the calculation by replace the ratio with weight, that is currentRatio (since totalRatio is identical for all pods)